src/neo/Network/P2P/TaskManager.cs

```diff
@@ -14,105 +14,140 @@ namespace Neo.Network.P2P
         private class Timer { }
 
         private static readonly TimeSpan TimerInterval = TimeSpan.FromSeconds(30);
         private static readonly TimeSpan TaskTimeout = TimeSpan.FromMinutes(1);
 
-        private readonly NeoSystem system;
         private const int MaxConncurrentTasks = 3;
-
+        private const int MaxSyncTasksCount = 50;
         private const int PingCoolingOffPeriod = 60; // in secconds.
+
+        private readonly NeoSystem system;
         /// <summary>
         /// A set of known hashes, of inventories or payloads, already received.
-        /// </summary>        
+        /// </summary>
         private readonly HashSetCache<UInt256> knownHashes;
         private readonly Dictionary<UInt256, int> globalTasks = new Dictionary<UInt256, int>();
+        private readonly Dictionary<uint, TaskSession> receivedBlockIndex = new Dictionary<uint, TaskSession>();
+        private readonly List<uint> failedSyncTasks = new List<uint>();
         private readonly Dictionary<IActorRef, TaskSession> sessions = new Dictionary<IActorRef, TaskSession>();
         private readonly ICancelable timer = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(TimerInterval, TimerInterval, Context.Self, new Timer(), ActorRefs.NoSender);
-
-        private readonly UInt256 HeaderTaskHash = UInt256.Zero;
-        private bool HasHeaderTask => globalTasks.ContainsKey(HeaderTaskHash);
+        private uint lastTaskIndex = 0;
 
         public TaskManager(NeoSystem system)
         {
             this.system = system;
             this.knownHashes = new HashSetCache<UInt256>(Blockchain.Singleton.MemPool.Capacity * 2 / 5);
+            this.lastTaskIndex = Blockchain.Singleton.Height;
+            Context.System.EventStream.Subscribe(Self, typeof(Blockchain.PersistCompleted));
         }
 
-        private void OnHeaderTaskCompleted()
+        private bool AssignSyncTask(uint index, TaskSession filterSession = null)
         {
-            if (!sessions.TryGetValue(Sender, out TaskSession session))
-                return;
-            session.Tasks.Remove(HeaderTaskHash);
-            DecrementGlobalTask(HeaderTaskHash);
-            RequestTasks(session);
+            if (index <= Blockchain.Singleton.Height || sessions.Values.Any(p => p != filterSession && p.IndexTasks.ContainsKey(index)))
+                return true;
+            Random rand = new Random();
+            KeyValuePair<IActorRef, TaskSession> remoteNode = sessions.Where(p => p.Value != filterSession && p.Value.LastBlockIndex >= index)
+                .OrderBy(p => p.Value.IndexTasks.Count)
+                .ThenBy(s => rand.Next())
+                .FirstOrDefault();
+            if (remoteNode.Value == null)
+            {
+                failedSyncTasks.Add(index);
+                return false;
+            }
+            TaskSession session = remoteNode.Value;
+            session.IndexTasks.Add(index, TimeProvider.Current.UtcNow);
```

Comment:
What happens to the local variable `session` here?

Reply:
It records the time that the task is assigned so that it can detect whether the task has timed out in the timer.